File: topk-py/tests/test_query_topk_formulas.py
import pytest
from topk_sdk import error
from topk_sdk.query import field, filter, select, fn

from . import ProjectContext
from .utils import dataset

def test_query_topk_clamping(ctx: ProjectContext):
    collection = dataset.books.setup(ctx)

    with pytest.raises(error.InvalidArgumentError):
        ctx.client.collection(collection.name).query(
            select(
                summary_distance=fn.vector_distance("summary_embedding", [2.0, 16]),
                bm25_score=fn.bm25_score()
            ).topk(
                (field("bm25_score").max(3).min(10)) 
                + (field("summary_distance") * 0.5),
                2,
                True,
            ),
            None,
            None
        )


def test_query_topk_pow_sqrt(ctx: ProjectContext):
    collection = dataset.books.setup(ctx)

    with pytest.raises(error.InvalidArgumentError):
        ctx.client.collection(collection.name).query(
        select(
            summary_distance=fn.vector_distance("summary_embedding", [2.0, 16]),
            bm25_score=fn.bm25_score()
        ).topk(
            (
                (field("bm25_score")**1.5) +
                pow(field("summary_distance"), 2)
            ).sqrt(),
            2,
            True,
        ),
        None,
        None
    )

def test_query_topk_exp_precision(ctx: ProjectContext):
    collection = dataset.books.setup(ctx)

    with pytest.raises(error.InvalidArgumentError):
        ctx.client.collection(collection.name).query(
            filter(
                abs(field("published_year").exp().ln() - 1988) < 10e-6
            ).topk(
                field("published_year"),
                2,
                True,
            ),
            None,
            None
        )

# python specific: modulo pow should fail
def test_modulo_pow_fail():
    with pytest.raises(NotImplementedError):
        pow(field("test"), 1.5, 5)

File: topk-rs/tests/test_query_topk_formulas.rs
use test_context::test_context;
use topk_rs::data::f32_vector;
use topk_rs::query::{field, filter, fns, select};

mod utils;
use utils::dataset;
use utils::ProjectTestContext;

use topk_rs::Error;

#[test_context(ProjectTestContext)]
#[tokio::test]
async fn test_query_topk_clamping(ctx: &mut ProjectTestContext) {
    let collection = dataset::books::setup(ctx).await;

    let err = ctx
        .client
        .collection(&collection.name)
        .query(
            select([
                (
                    "summary_distance",
                    fns::vector_distance("summary_embedding", f32_vector(vec![2.0; 16])),
                ),
                ("bm25_score", fns::bm25_score()),
            ])
            .topk(
                field("bm25_score")
                    .max(3)
                    .min(10)
                    .add(field("summary_distance").mul(0.5)),
                2,
                true,
            ),
            None,
            None,
        )
        .await
        .expect_err("max, min not implemented yet");

    assert!(matches!(err, Error::InvalidArgument(_)));
}

#[test_context(ProjectTestContext)]
#[tokio::test]
async fn test_query_topk_pow_sqrt(ctx: &mut ProjectTestContext) {
    let collection = dataset::books::setup(ctx).await;

    let err = ctx
        .client
        .collection(&collection.name)
        .query(
            select([
                (
                    "summary_distance",
                    fns::vector_distance("summary_embedding", f32_vector(vec![2.0; 16])),
                ),
                ("bm25_score", fns::bm25_score()),
            ])
            .topk(
                field("bm25_score")
                    .pow(1.5)
                    .add(field("summary_distance").pow(2))
                    .sqrt(),
                2,
                true,
            ),
            None,
            None,
        )
        .await
        .expect_err("pow, sqrt not implemented yet");

    assert!(matches!(err, Error::InvalidArgument(_)));
}

#[test_context(ProjectTestContext)]
#[tokio::test]
async fn test_query_topk_exp_precision(ctx: &mut ProjectTestContext) {
    let collection = dataset::books::setup(ctx).await;

    let err = ctx
        .client
        .collection(&collection.name)
        .query(
            filter(
                field("published_year")
                    .exp()
                    .ln()
                    .sub(1988 as u32)
                    .abs()
                    .lte(10e-6 as u32),
            )
            .topk(field("published_year"), 2, true),
            None,
            None,
        )
        .await
        .expect_err("exp, ln, abs not implemented yet");

    assert!(matches!(err, Error::InvalidArgument(_)));
    // assert_doc_ids!(result, ["alchemist"]);
}
